ios/RNCameraRoll.m

```diff
@@ -60,86 +85,96 @@ @implementation RNCameraRoll
           // if creation timestamp is less than our filter and the image has never been modified, return
           return;
-        } else if ( [extension caseInsensitiveCompare:@"heic"] == NSOrderedSame ) {
-          [self _processHEIC:asset orgFilename:orgFilename];
         } else {
-          [self _processJPG:asset orgFilename:orgFilename];
+          dispatch_group_enter(assetDatasGroup);
+          [self _writeToDisk:asset orgFilename:orgFilename onComplete:^(NSString *path, UIImageOrientation orientation) {
+            if (path) {
+              AssetData *data = [[AssetData alloc] init];
+              data.asset = asset;
+              data.path = path;
+              data.orientation = orientation;
+              [assetDatas addObject:data];
+            }
+            dispatch_group_leave(assetDatasGroup);
+          }];
         }
       }];
+
+      // assetDatas fully populated now
+      [assetDatas sortUsingSelector:@selector(compare:)];
+
+      NSMutableArray<NSDictionary*> *results = [[NSMutableArray alloc] initWithCapacity:assetDatas.count];
+
+      [assetDatas enumerateObjectsUsingBlock:^(AssetData *assetData, NSUInteger idx, BOOL *stop) {
+        [results addObject:[self _convertAssetData:assetData]];
+      }];
+
+      NSError *serializationError;
+      NSData *data = [NSJSONSerialization dataWithJSONObject:results options:NSJSONWritingPrettyPrinted error:&serializationError];
+      if(serializationError) {
+        reject(@"0", serializationError.localizedFailureReason, serializationError);
+      } else {
+        NSString* jsonStr = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
+        resolve(jsonStr);
+      }
     }
+  } else {
+    reject(@"1", @"no photos permission", [[NSError alloc] initWithDomain:@"io.textile" code:1 userInfo:nil]);
   }
-  resolve(@YES);
 }
 
-- (void)_processJPG:(PHAsset *)asset orgFilename:(NSString *)orgFilename  {
+- (void)_writeToDisk:(PHAsset *)asset orgFilename:(NSString *)orgFilename onComplete:(void (^_Nonnull)(NSString*, UIImageOrientation))onComplete  {
   // If the file isn't a HEIC, just write it to temp and move along.
-  [[PHImageManager defaultManager] requestImageDataForAsset:asset options:nil resultHandler:^(NSData * _Nullable imageData, NSString * _Nullable dataUTI, UIImageOrientation imageOrientation, NSDictionary * _Nullable info) {
+  PHImageRequestOptions *requestOptions = [[PHImageRequestOptions alloc] init];
+//  requestOptions.synchronous = @TRUE;
+  [[PHImageManager defaultManager] requestImageDataForAsset:asset options:requestOptions resultHandler:^(NSData * _Nullable imageData, NSString * _Nullable dataUTI, UIImageOrientation imageOrientation, NSDictionary * _Nullable info) {
     // Got the image data for copying
     if (imageData) {
+      UIImage *image = [UIImage imageWithData:imageData];
+      NSData *jpegData = UIImageJPEGRepresentation(image, 1.0);
```

Comment:
No matter if the image originally was hiec or jpeg, we create a `UIImage` from it and then jpeg `NSData`.